Fails to build with `python setup.py build`

Unlike other Python packages, this one fails to build:
```
  running build
  running build_py
  creating build
  creating build/lib.mingw_x86_64_clang-cpython-311
  creating build/lib.mingw_x86_64_clang-cpython-311/pybase64
  copying src/pybase64/_fallback.py -> build/lib.mingw_x86_64_clang-cpython-311/pybase64
  copying src/pybase64/_license.py -> build/lib.mingw_x86_64_clang-cpython-311/pybase64
  copying src/pybase64/_version.py -> build/lib.mingw_x86_64_clang-cpython-311/pybase64
  copying src/pybase64/__init__.py -> build/lib.mingw_x86_64_clang-cpython-311/pybase64
  copying src/pybase64/__main__.py -> build/lib.mingw_x86_64_clang-cpython-311/pybase64
  copying src/pybase64/py.typed -> build/lib.mingw_x86_64_clang-cpython-311/pybase64
  running build_ext
  running cmake '--version'
  cmake version 3.28.1
  
  CMake suite maintained and supported by Kitware (kitware.com/cmake).
  getting cmake extra config
    machine: amd64
    platform: mingw_x86_64_clang
    plat_name: mingw_x86_64_clang
    ARCHFLAGS: None
    CC: clang
    CFLAGS: -march=nocona -msahf -mtune=generic -O2 -pipe -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector-strong
    LDFLAGS: -pipe
    sysconfig CC: clang
    sysconfig CCSHARED: 
    sysconfig CFLAGS: -DNDEBUG -g -fwrapv -O3 -Wall -march=nocona -msahf -mtune=generic -O2 -pipe -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector-strong -O3 -march=nocona -msahf -mtune=generic -O2 -pipe -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector-strong -O3
    sysconfig BASECFLAGS: 
    sysconfig OPT: -DNDEBUG -g -fwrapv -O3 -Wall
    sysconfig LDFLAGS: -pipe -pipe
  building 'pybase64._pybase64' extension
  creating build/temp.mingw_x86_64_clang-cpython-311
  creating build/temp.mingw_x86_64_clang-cpython-311/src
  creating build/temp.mingw_x86_64_clang-cpython-311/src/pybase64
  clang -DNDEBUG -g -fwrapv -O3 -Wall -march=nocona -msahf -mtune=generic -O2 -pipe -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector-strong -O3 -march=nocona -msahf -mtune=generic -O2 -pipe -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector-strong -O3 -march=nocona -msahf -mtune=generic -O2 -pipe -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector-strong -D__USE_MINGW_ANSI_STDIO=1 -DBASE64_STATIC_DEFINE=1 -Ibase64/include/ -Ibase64/lib/ -I.base64_build -ID:/M/msys64/clang64/include/python3.11 -c src/pybase64/_pybase64.c -o build/temp.mingw_x86_64_clang-cpython-311/src/pybase64/_pybase64.o
  src/pybase64/_pybase64.c:4:10: fatal error: 'config.h' file not found
      4 | #include <config.h>
        |          ^~~~~~~~~~
  1 error generated.
  warning: BuildExt: building extension "pybase64._pybase64" failed: command 'D:\\M\\msys64\\clang64\\bin/clang.exe' failed with exit code 1
```
It expects base64 to be in the directory (`-Ibase64/include/`) but setuptools creates a temporary build directory.

--- a/.github/workflows/msys2.yml
+++ b/.github/workflows/msys2.yml
@@ -1,0 +1,46 @@
+name: MSYS2 Tests
+
+on:
+  push:
+    branches-ignore:
+      - "dependabot/**"
+  pull_request:
+
+jobs:
+  test:
+    name: Test with MSYS2 ${{ matrix.sys }}
+    runs-on: windows-2022
+    strategy:
+      matrix:
+        include:
+          - { sys: msys, toolchain: "gcc" }
+          - { sys: mingw64, env: mingw-w64-x86_64- }
+          - { sys: ucrt64, env: mingw-w64-ucrt-x86_64- }
+          - { sys: clang64, env: mingw-w64-clang-x86_64- }
+    steps:
+    - uses: actions/checkout@v4
+      with:
+        submodules: recursive
+    - name: 'Setup MSYS2'
+      uses: msys2/setup-msys2@v2
+      with:
+        msystem: ${{matrix.sys}}
+        update: true
+        install: >-
+          make
+          ${{matrix.env}}cmake
+          ${{matrix.env}}ninja
+          ${{matrix.env}}${{matrix.toolchain || 'toolchain' }}
+          ${{matrix.env}}python
+          ${{matrix.env}}python-pip
+          ${{matrix.env}}python-pytest
+          ${{matrix.env}}python-setuptools
+          ${{matrix.env}}python-typing_extensions
+          ${{matrix.env}}python-wheel
+    - name: "Run tests"
+      shell: msys2 {0}
+      env:
+        CIBUILDWHEEL: 1
+      run: |
+        python -m pip install -v --no-build-isolation .
+        python -m pytest

--- a/README.rst
+++ b/README.rst
@@ -123,6 +123,8 @@ Changelog
 =========
 Future
 ------
+- Add MSYS2 support for C-extension
+- Better logging on base64 build failure when C-extension build is optional
 - Drop python 3.6 support
 
 1.3.2

--- a/setup.py
+++ b/setup.py
@@ -58,7 +58,7 @@ pybase64_ext = Extension(
 )
 
 
-def get_cmake_extra_config(plat_name: str | None, build_type: str) -> list[str]:
+def get_cmake_extra_config(plat_name: str | None, build_type: str) -> tuple[bool, list[str]]:
     log.info("getting cmake extra config")
     extra_config = []
     machine = platform_module.machine().lower()
@@ -80,13 +80,14 @@ def get_cmake_extra_config(plat_name: str | None, build_type: str) -> list[str]:
     log.info("  sysconfig LDFLAGS: %s", sysconfig.get_config_var("LDFLAGS"))
 
     platform = plat_name or platform
+    is_msvc = platform.startswith("win")
 
-    if not IS_WINDOWS:
+    if not is_msvc:
         extra_config.append(f"-DCMAKE_BUILD_TYPE={build_type}")
 
-    if IS_WINDOWS:
-        if not platform.startswith("win"):
-            msg = f"Building {platform} is not supported on Windows"
+    if is_msvc:
+        if not IS_WINDOWS:
+            msg = f"Building {platform} is only supported on Windows"
             raise ValueError(msg)
         # setup cross-compile
         # assumes VS2019 or VS2022 will be used as the default generator
@@ -129,7 +130,7 @@ def get_cmake_extra_config(plat_name: str | None, build_type: str) -> list[str]:
         else:
             log.warning("`%s` is not a known value for CMAKE_OSX_ARCHITECTURES", arch)
 
-    return extra_config
+    return is_msvc, extra_config
 
 
 def cmake(*args: str) -> None:
@@ -158,10 +159,11 @@ def base64_build(plat_name: str | None) -> Generator[bool, None, None]:
     try:
         try:
             cmake("--version")
-            config_options.extend(get_cmake_extra_config(plat_name, build_type))
+            is_msvc, extra_config = get_cmake_extra_config(plat_name, build_type)
+            config_options.extend(extra_config)
             cmake(*config_options)
             cmake("--build", str(build_dir), "--config", build_type, "--verbose")
-            if IS_WINDOWS:
+            if is_msvc:
                 shutil.copyfile(build_dir / build_type / "base64.lib", build_dir / "base64.lib")
             base64_built = True
         except Exception as e:

--- a/setup.py
+++ b/setup.py
@@ -139,7 +139,8 @@ def cmake(*args: str) -> None:
 
 
 @contextmanager
-def base64_build(plat_name: str | None) -> Generator[None, None, None]:
+def base64_build(plat_name: str | None) -> Generator[bool, None, None]:
+    base64_built = False
     source_dir = HERE / "base64"
     build_dir = HERE / ".base64_build"
     build_type = "Release"
@@ -162,14 +163,15 @@ def base64_build(plat_name: str | None) -> Generator[None, None, None]:
             cmake("--build", str(build_dir), "--config", build_type, "--verbose")
             if IS_WINDOWS:
                 shutil.copyfile(build_dir / build_type / "base64.lib", build_dir / "base64.lib")
-        except Exception:
+            base64_built = True
+        except Exception as e:
+            log.error("error: %s", e)
             if not OPTIONAL_EXTENSION:
                 raise
-        yield
+        yield base64_built
     finally:
-        # if build_dir.exists():
-        #     shutil.rmtree(build_dir)
-        pass
+        if build_dir.exists():
+            shutil.rmtree(build_dir, ignore_errors=True)
 
 
 class BuildExt(build_ext):
@@ -185,8 +187,12 @@ class BuildExt(build_ext):
 
     def run(self) -> None:
         plat_name = getattr(self, "plat_name", None)
-        with base64_build(plat_name):
-            super().run()
+        with base64_build(plat_name) as build_successful:
+            if build_successful:
+                super().run()
+            else:
+                assert OPTIONAL_EXTENSION
+                log.warning("warning: skipping optional C-extension, base64 library build failed")
 
 
 setup(

